edfi-repo-auditor/edfi_repo_auditor/auditor.py

```diff
@@ -0,0 +1,163 @@
+# SPDX-License-Identifier: Apache-2.0
+# Licensed to the Ed-Fi Alliance under one or more agreements.
+# The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+# See the LICENSE and NOTICES files in the project root for more information.
+
+import json
+import logging
+import re
+import time
+
+from edfi_repo_auditor.config import Configuration
+from edfi_repo_auditor.github_client import GitHubClient
+from datetime import datetime, timedelta
+from pprint import pformat
+
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+# Parameters to evaluate dependabot alerts
+ALERTS_INCLUDED_SEVERITIES = ["CRITICAL", "HIGH"]
+ALERTS_WEEKS_SINCE_CREATED = 3
+
+
+def run_audit(config: Configuration) -> None:
+    start = time.time()
+    client = GitHubClient(config.personal_access_token)
+
+    repositories = config.repositories if len(config.repositories) > 0 else client.get_repositories(config.organization)
+
+    report: dict = {}
+    for repo in repositories:
+        repo_config = get_repo_information(client, config.organization, repo)
+        logger.debug(f"Repo configuration: {repo_config}")
+        actions = audit_actions(client, config.organization, repo)
+        logger.debug(f"Actions {actions}")
+        file_review = review_files(client, config.organization, repo)
+        logger.debug(f"Files: {file_review}")
+
+        checklist = actions | file_review | repo_config
+        auditing_rules = get_file()
+        score = get_result(checklist, auditing_rules["rules"])
+        logger.debug(f"Rules to follow: {auditing_rules}")
+
+        report[repo] = {
+            "score": score,
+            "result": "OK" if score > auditing_rules["threshold"] else "Action required",
+            "description": checklist
+        }
+
+    if config.save_results is True:
+        logger.info("Saving report to reports/audit-result.json")
+        json_report = json.dumps(report, indent=4)
+
+        with open("reports/audit-result.json", "w") as outfile:
+            outfile.write(json_report)
+    else:
+        logger.info(pformat(report))
+
+    logger.info(f"Finished auditing repositories for {config.organization} in {'{:.2f}'.format(time.time() - start)} seconds")
+
+
+def audit_actions(client: GitHubClient, organization: str, repository: str) -> dict:
+    audit_results = {
+        "Has Actions": False,
+        "Uses CodeQL": False,
+        "Uses Allowed list": False,
```

Comment:
I couldn't remember what "Allowed list" means when I read this.

```suggestion
        "Uses only approved GitHub Actions": False,
```

Reply:
Probably want to extract that to a constant so that you can reuse that constant on lines 88 and 89. Also extract the other string labels to constants.

Reply:
I extracted this to a namedtuple, still need to work on the property names and values to be clearer. I will continue to work on this while also adding the changes requested by Vinaya of using other messages instead of True and False in the results